packages/model-viewer/src/three-components/Renderer.ts

```diff
@@ -169,46 +182,81 @@ export class Renderer extends EventDispatcher {
       height = Math.max(height, scene.height);
     }
 
-    this.threeRenderer.getSize(size);
-    if (width === size.x && height === size.y && dprUpdated === false) {
+    if (width === this.width && height === this.height && dpr === this.dpr) {
       return;
     }
+    this.width = width;
+    this.height = height;
+    this.dpr = dpr;
 
-    // The canvas element must by styled outside of three due to the offscreen
-    // canvas not being directly stylable.
-    this.canvasElement.style.width = `${width}px`;
-    this.canvasElement.style.height = `${height}px`;
     if (this.canRender) {
-      this.threeRenderer.setSize(width, height, false);
+      this.threeRenderer.setSize(width * dpr, height * dpr, false);
     }
 
+    // Expand the canvas size to make up for shrinking the viewport.
+    const widthCSS = width / this.scale;
+    const heightCSS = height / this.scale;
+    // The canvas element must by styled outside of three due to the offscreen
+    // canvas not being directly stylable.
+    this.canvasElement.style.width = `${widthCSS}px`;
+    this.canvasElement.style.height = `${heightCSS}px`;
+
     // Each scene's canvas must match the renderer size. In general they can be
     // larger than the element that contains them, but the overflow is hidden
     // and only the portion that is shown is copied over.
     for (const scene of this.scenes) {
       const {canvas} = scene;
-      const {width: pixelWidth, height: pixelHeight} =
-          this.threeRenderer.domElement;
-      canvas.width = pixelWidth;
-      canvas.height = pixelHeight;
-      canvas.style.width = `${width}px`;
-      canvas.style.height = `${height}px`;
+      canvas.width = width * dpr;
+      canvas.height = height * dpr;
+      canvas.style.width = `${widthCSS}px`;
+      canvas.style.height = `${heightCSS}px`;
       scene.isDirty = true;
     }
   }
 
+  private updateRendererScale() {
+    let {scale} = this;
+    if (this.avgFrameDuration > HIGH_FRAME_DURATION && scale > MIN_SCALE) {
+      scale *= SCALE_STEP;
+      scale = Math.max(scale, MIN_SCALE);
+    } else if (this.avgFrameDuration < LOW_FRAME_DURATION && scale < 1) {
+      scale /= SCALE_STEP;
+      scale = Math.min(scale, 1);
+    }
+
+    if (scale == this.scale) {
+      return;
+    }
+    this.scale = scale;
+    this.avgFrameDuration = (HIGH_FRAME_DURATION + LOW_FRAME_DURATION) / 2;
+
+    const width = this.width / scale;
+    const height = this.height / scale;
+
+    this.canvasElement.style.width = `${width}px`;
+    this.canvasElement.style.height = `${height}px`;
+    for (const scene of this.scenes) {
+      const {style} = scene.canvas;
+      style.width = `${width}px`;
+      style.height = `${height}px`;
+      scene.isDirty = true;
+    }
+    console.log('scale = ', this.scale);
```

Comment:
TODO: remove this before merge; Again this helps with seeing what the dynamic resizing is doing.